File: app/src/main/java/com/example/stash/convert/ConversionEngine.kt
package com.example.stash.convert

import com.example.stash.download.DownloadFormat
import com.example.stash.download.DownloadQuality
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import java.io.File

/**
 * Audio conversion engine wrapping FFmpeg via ProcessBuilder.
 *
 * Converts raw downloaded audio files (typically .webm, .opus, .m4a from yt-dlp)
 * to the user's selected format (MP3/AAC) at the user's selected quality (bitrate).
 *
 * Processes one file at a time sequentially.
 */
class ConversionEngine {

    /**
     * Converts a raw audio file to the specified format and quality.
     *
     * @param inputPath Absolute path to the raw downloaded audio file.
     * @param format Target output format (MP3 or AAC).
     * @param quality Target audio quality (bitrate).
     * @param onProgress Optional callback for conversion progress (0.0 to 1.0).
     * @return Absolute path to the converted output file.
     * @throws ConversionException if conversion fails.
     */
    suspend fun convert(
        inputPath: String,
        format: DownloadFormat,
        quality: DownloadQuality,
        onProgress: ((Float) -> Unit)? = null
    ): String = withContext(Dispatchers.IO) {
        val inputFile = File(inputPath)
        if (!inputFile.exists()) {
            throw ConversionException("Input file not found: $inputPath")
        }

        // Build output path: same directory, same base name, new extension
        val outputFile = File(
            inputFile.parent,
            "${inputFile.nameWithoutExtension}.${format.extension}"
        )

        // If the input is already in the correct format, check if we just need to re-encode for quality
        val inputExtension = inputFile.extension.lowercase()
        if (inputExtension == format.extension) {
            // Already correct format — just rename/keep
            println("File already in ${format.extension} format, skipping conversion: ${inputFile.name}")
            return@withContext inputPath
        }

        println("Converting: ${inputFile.name} → ${outputFile.name} (${format.ffmpegCodec} @ ${quality.bitrateKbps}kbps)")

        // First, get the duration of the input file for progress tracking
        val durationSeconds = getDuration(inputPath)

        val cmd = mutableListOf(
            "ffmpeg",
            "-i", inputPath,
            "-y",                          // Overwrite output
            "-vn",                         // No video
            "-codec:a", format.ffmpegCodec,
            "-b:a", "${quality.bitrateKbps}k",
            "-ar", "44100",               // Standard sample rate
            "-ac", "2",                    // Stereo
            "-progress", "pipe:1",         // Output progress to stdout
            "-loglevel", "error",          // Only show errors on stderr
            outputFile.absolutePath
        )

        try {
            val process = ProcessBuilder(cmd)
                .redirectErrorStream(false)
                .start()

            // Read progress from stdout
            process.inputStream.bufferedReader().useLines { lines ->
                lines.forEach { line ->
                    if (line.startsWith("out_time_us=")) {
                        val timeUs = line.substringAfter("out_time_us=").toLongOrNull()
                        if (timeUs != null && durationSeconds > 0) {
                            val currentSeconds = timeUs / 1_000_000.0
                            val progress = (currentSeconds / durationSeconds).toFloat().coerceIn(0f, 1f)
                            onProgress?.invoke(progress)
                        }
                    }
                }
            }

            // Also drain stderr to prevent blocking
            val stderrOutput = process.errorStream.bufferedReader().readText()

            val exitCode = process.waitFor()
            if (exitCode != 0) {
                throw ConversionException(
                    "FFmpeg exited with code $exitCode: ${stderrOutput.take(500)}"
                )
            }

            if (!outputFile.exists()) {
                throw ConversionException("Conversion completed but output file not found: ${outputFile.absolutePath}")
            }

            // Delete the raw input file after successful conversion
            if (inputFile.absolutePath != outputFile.absolutePath) {
                inputFile.delete()
                println("Deleted raw file: ${inputFile.name}")
            }

            println("Conversion complete: ${outputFile.name} (${outputFile.length() / 1024}KB)")
            onProgress?.invoke(1f)

            return@withContext outputFile.absolutePath

        } catch (e: ConversionException) {
            throw e
        } catch (e: Exception) {
            System.err.println("Conversion failed: ${e.message}")
            e.printStackTrace()
            throw ConversionException("Conversion failed: ${e.message}", e)
        }
    }

    /**
     * Gets the duration of an audio file in seconds using ffprobe.
     * Returns 0 if duration cannot be determined.
     */
    private fun getDuration(filePath: String): Double {
        return try {
            val cmd = listOf(
                "ffprobe",
                "-v", "quiet",
                "-show_entries", "format=duration",
                "-of", "default=noprint_wrappers=1:nokey=1",
                filePath
            )

            val process = ProcessBuilder(cmd)
                .redirectErrorStream(true)
                .start()

            val output = process.inputStream.bufferedReader().readText().trim()
            process.waitFor()

            output.toDoubleOrNull() ?: 0.0
        } catch (e: Exception) {
            println("Could not determine duration: ${e.message}")
            0.0
        }
    }
}

class ConversionException(message: String, cause: Throwable? = null) : Exception(message, cause)

File: src/main/services/ConversionEngine.ts
import { spawn, execFile, ChildProcess } from 'child_process'
import path from 'path'
import fs from 'fs'
import readline from 'readline'
import { DownloadFormat, DownloadQuality } from '../../shared/types'
import { DependencyResolver } from './DependencyResolver'

export class ConversionEngine {
  private activeProcesses: Map<string, ChildProcess> = new Map()

  /**
   * Converts a downloaded audio/video file to the desired output format & bitrate.
   */
  public async convert(
    inputPath: string,
    format: DownloadFormat,
    quality: DownloadQuality,
    conversionId: string,
    onProgress?: (progressFraction: number) => void
  ): Promise<string> {
    if (!fs.existsSync(inputPath)) {
      throw new Error(`Input file not found: ${inputPath}`)
    }

    const ext = this.getFormatExtension(format)
    const inputParsed = path.parse(inputPath)
    const outputPath = path.join(inputParsed.dir, `${inputParsed.name}.${ext}`)

    // If input already has target format and doesn't need re-encoding
    if (inputParsed.ext.replace('.', '').toLowerCase() === ext.toLowerCase()) {
      return inputPath
    }

    const durationSeconds = await this.getDuration(inputPath)
    const ffmpegPath = DependencyResolver.resolveExecutable('ffmpeg')

    const args: string[] = ['-i', inputPath, '-y']

    if (format === 'MP4' || format === 'OTHER_VIDEO') {
      args.push('-c:v', 'copy', '-c:a', 'aac', '-b:a', '192k')
    } else {
      args.push('-vn') // No video
      const { codec, bitrate } = this.getAudioCodecAndBitrate(format, quality)
      args.push('-codec:a', codec)
      if (bitrate) {
        args.push('-b:a', bitrate)
      }
      args.push('-ar', '44100', '-ac', '2')
    }

    args.push('-progress', 'pipe:1', '-loglevel', 'error', outputPath)

    return new Promise<string>((resolve, reject) => {
      const child = spawn(ffmpegPath, args, { windowsHide: true })
      this.activeProcesses.set(conversionId, child)

      const rl = readline.createInterface({ input: child.stdout })

      rl.on('line', (line) => {
        if (line.startsWith('out_time_us=')) {
          const timeUs = parseInt(line.replace('out_time_us=', '').trim(), 10)
          if (!isNaN(timeUs) && durationSeconds > 0) {
            const currentSeconds = timeUs / 1000000.0
            const fraction = Math.min(Math.max(currentSeconds / durationSeconds, 0), 1)
            if (onProgress) {
              onProgress(fraction)
            }
          }
        }
      })

      let stderrOutput = ''
      child.stderr.on('data', (data) => {
        stderrOutput += data.toString()
      })

      child.on('close', (code) => {
        this.activeProcesses.delete(conversionId)
        if (code === 0) {
          if (fs.existsSync(outputPath)) {
            // Delete raw input file if different
            if (inputPath !== outputPath) {
              try {
                fs.unlinkSync(inputPath)
              } catch (e) {
                // ignore
              }
            }
            if (onProgress) onProgress(1)
            resolve(outputPath)
          } else {
            reject(new Error('Conversion finished but target file missing'))
          }
        } else {
          reject(new Error(`FFmpeg error (code ${code}): ${stderrOutput.slice(-300)}`))
        }
      })

      child.on('error', (err) => {
        this.activeProcesses.delete(conversionId)
        reject(new Error(`FFmpeg failed to start: ${err.message}`))
      })
    })
  }

  /**
   * Cancels an active conversion
   */
  public cancelConversion(conversionId: string): void {
    const proc = this.activeProcesses.get(conversionId)
    if (proc && !proc.killed) {
      try {
        proc.kill('SIGTERM')
      } catch (e) {
        try {
          proc.kill('SIGKILL')
        } catch (err) {
          // ignore
        }
      }
      this.activeProcesses.delete(conversionId)
    }
  }

  private getFormatExtension(format: DownloadFormat): string {
    switch (format) {
      case 'MP3': return 'mp3'
      case 'AAC': return 'm4a'
      case 'FLAC': return 'flac'
      case 'OPUS': return 'opus'
      case 'WAV': return 'wav'
      case 'MP4':
      case 'OTHER_VIDEO': return 'mp4'
      case 'AUTO':
      default: return 'mp3'
    }
  }

  private getAudioCodecAndBitrate(format: DownloadFormat, quality: DownloadQuality): { codec: string; bitrate?: string } {
    const isHighOrAbove = quality === '4K' || quality === '2K' || quality === 'HIGH'
    switch (format) {
      case 'MP3': {
        const kbps = isHighOrAbove ? '320k' : (quality === 'MID' ? '192k' : '128k')
        return { codec: 'libmp3lame', bitrate: kbps }
      }
      case 'AAC': {
        const kbps = isHighOrAbove ? '256k' : (quality === 'MID' ? '192k' : '128k')
        return { codec: 'aac', bitrate: kbps }
      }
      case 'FLAC':
        return { codec: 'flac' }
      case 'OPUS': {
        const kbps = isHighOrAbove ? '256k' : (quality === 'MID' ? '160k' : '96k')
        return { codec: 'libopus', bitrate: kbps }
      }
      case 'WAV':
        return { codec: 'pcm_s16le' }
      case 'AUTO':
      default: {
        const kbps = isHighOrAbove ? '320k' : (quality === 'MID' ? '192k' : '128k')
        return { codec: 'libmp3lame', bitrate: kbps }
      }
    }
  }

  private getDuration(filePath: string): Promise<number> {
    return new Promise((resolve) => {
      const ffprobePath = DependencyResolver.resolveExecutable('ffprobe')
      const args = [
        '-v', 'quiet',
        '-show_entries', 'format=duration',
        '-of', 'default=noprint_wrappers=1:nokey=1',
        filePath
      ]
      execFile(ffprobePath, args, { timeout: 10000 }, (err, stdout) => {
        if (!err && stdout) {
          const duration = parseFloat(stdout.trim())
          resolve(!isNaN(duration) ? duration : 0)
        } else {
          resolve(0)
        }
      })
    })
  }
}
